File: scaffold/logging.py
"""Experiment-based logging and summary generation."""

import json
import re
from pathlib import Path
from datetime import datetime
from dataclasses import dataclass, field
from typing import Dict, List, Any, Optional, Callable

PROJECT_ROOT = Path(__file__).parent.parent
LOGS_DIR = PROJECT_ROOT / "logs"

# Regex to strip ANSI escape codes
ANSI_ESCAPE = re.compile(r'\x1B(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])')


def strip_ansi(text: str) -> str:
    """Remove ANSI escape codes from text."""
    return ANSI_ESCAPE.sub('', text)


# =============================================================================
# OUTPUT PARSING
# =============================================================================

def parse_output(stdout: str) -> Dict[str, Any]:
    """Parse stream-json output into structured data."""
    if not stdout:
        return {"messages": []}
    messages = []
    for line in stdout.strip().split('\n'):
        try:
            messages.append(json.loads(line))
        except json.JSONDecodeError:
            pass
    return {"messages": messages}


def extract_events(parsed: Dict[str, Any]) -> Dict[str, Any]:
    """Extract events (tool calls, files, etc.) from parsed output."""
    events = {
        "tool_calls": [], "files_read": [], "files_created": [],
        "files_modified": [], "commands_run": [], "skills_invoked": [],
        "duration_seconds": None, "num_turns": None,
    }

    # Map tool_use_id -> index in tool_calls list for matching outputs
    tool_id_to_index = {}

    for msg in parsed.get("messages", []):
        if msg.get("type") == "result":
            events["duration_seconds"] = msg.get("duration_ms", 0) / 1000
            events["num_turns"] = msg.get("num_turns")

        if msg.get("type") == "assistant":
            for item in msg.get("message", {}).get("content", []):
                if item.get("type") == "tool_use":
                    tool, inp = item.get("name", ""), item.get("input", {})
                    tool_id = item.get("id")
                    tool_call = {"tool": tool, "input": inp}
                    if tool_id:
                        tool_id_to_index[tool_id] = len(events["tool_calls"])
                    events["tool_calls"].append(tool_call)
                    path = inp.get("file_path", "")
                    if tool == "Read" and path:
                        events["files_read"].append(path)
                    elif tool == "Write" and path:
                        events["files_created"].append(path)
                    elif tool == "Edit" and path:
                        events["files_modified"].append(path)
                    elif tool == "Bash" and inp.get("command"):
                        events["commands_run"].append(inp["command"])
                    elif tool == "Skill" and inp.get("skill"):
                        events["skills_invoked"].append(inp["skill"])

        # Capture tool results and match to their tool_use calls
        if msg.get("type") == "user":
            for item in msg.get("message", {}).get("content", []):
                if item.get("type") == "tool_result":
                    tool_use_id = item.get("tool_use_id")
                    if tool_use_id and tool_use_id in tool_id_to_index:
                        idx = tool_id_to_index[tool_use_id]
                        # Extract output content
                        content = item.get("content", "")
                        if isinstance(content, list):
                            # Content can be a list of text blocks
                            content = " ".join(
                                c.get("text", str(c)) if isinstance(c, dict) else str(c)
                                for c in content
                            )
                        events["tool_calls"][idx]["output"] = content

    return events


# =============================================================================
# TREATMENT RESULT
# =============================================================================

@dataclass
class TreatmentResult:
    """Result from a single treatment run."""
    name: str
    passed: bool
    checks_passed: List[str]
    checks_failed: List[str]
    events_summary: Dict[str, Any] = field(default_factory=dict)

    def has_check(self, pattern: str) -> bool:
        """Check if any passed check contains pattern."""
        return any(pattern in c for c in self.checks_passed)

    def has_failed_check(self, pattern: str) -> bool:
        """Check if any failed check contains pattern."""
        return any(pattern in c for c in self.checks_failed)

    @property
    def turns(self) -> Optional[int]:
        return self.events_summary.get("num_turns")

    @property
    def duration(self) -> Optional[float]:
        return self.events_summary.get("duration_seconds")

    @property
    def tool_calls(self) -> Optional[int]:
        return self.events_summary.get("tool_calls")


# =============================================================================
# REPORT COLUMNS
# =============================================================================

@dataclass
class ReportColumn:
    """Defines a column in the results table."""
    name: str
    extract: Callable[[TreatmentResult], str]  # Single run -> display value
    aggregate: Callable[[List[TreatmentResult]], str] = None  # Multiple runs -> display value

    def get_value(self, result: TreatmentResult) -> str:
        return self.extract(result)

    def get_aggregate(self, runs: List[TreatmentResult]) -> str:
        if self.aggregate:
            return self.aggregate(runs)
        return self.extract(runs[0]) if runs else "N/A"


def bool_column(name: str, pattern: str) -> ReportColumn:
    """Column that checks if pattern exists in passed checks."""
    return ReportColumn(
        name=name,
        extract=lambda r: "Yes" if r.has_check(pattern) else "No",
        aggregate=lambda runs: f"{sum(1 for r in runs if r.has_check(pattern))}/{len(runs)}",
    )


def quality_column(name: str = "Quality") -> ReportColumn:
    """Column for output quality ([GOOD] vs [LOW])."""
    def extract(r):
        for c in r.checks_passed:
            if "[GOOD]" in c:
                return "Good"
            if "[LOW]" in c:
                return "Low"
        return "N/A"

    def aggregate(runs):
        good = sum(1 for r in runs if any("[GOOD]" in c for c in r.checks_passed))
        return f"{good}/{len(runs)}"

    return ReportColumn(name=name, extract=extract, aggregate=aggregate)


def default_columns() -> List[ReportColumn]:
    """Standard columns: Pass, Turns, Duration, Tools."""
    return [
        ReportColumn(
            name="Pass",
            extract=lambda r: "PASS" if r.passed else "FAIL",
            aggregate=lambda runs: f"{sum(1 for r in runs if r.passed)}/{len(runs)}",
        ),
        ReportColumn(
            name="Turns",
            extract=lambda r: str(r.turns) if r.turns else "N/A",
            aggregate=lambda runs: _avg([r.turns for r in runs if r.turns], "{:.0f}"),
        ),
        ReportColumn(
            name="Duration",
            extract=lambda r: f"{r.duration:.0f}s" if r.duration else "N/A",
            aggregate=lambda runs: _avg([r.duration for r in runs if r.duration], "{:.0f}s"),
        ),
        ReportColumn(
            name="Tools",
            extract=lambda r: str(r.tool_calls) if r.tool_calls else "N/A",
            aggregate=lambda runs: _avg([r.tool_calls for r in runs if r.tool_calls], "{:.0f}"),
        ),
    ]


def _avg(values: List, fmt: str = "{:.1f}") -> str:
    """Calculate average and format, or return N/A."""
    values = [v for v in values if v is not None]
    if not values:
        return "N/A"
    return fmt.format(sum(values) / len(values))


# =============================================================================
# EXPERIMENT LOGGER
# =============================================================================

class ExperimentLogger:
    """Manages logging for a single experiment run."""

    def __init__(self, experiment_name: str = None, columns: List[ReportColumn] = None):
        """Create experiment logger.

        Args:
            experiment_name: Name for this experiment
            columns: Custom columns for reporting (in addition to defaults)
        """
        self.timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        self.name = experiment_name or f"experiment_{self.timestamp}"
        self.experiment_id = f"{self.name}_{self.timestamp}"
        self.base_dir = LOGS_DIR / "experiments" / self.experiment_id

        # Create subdirectories
        self.code_dir = self.base_dir / "code"
        self.docker_dir = self.base_dir / "docker"
        self.events_dir = self.base_dir / "events"
        self.reports_dir = self.base_dir / "reports"
        self.raw_dir = self.base_dir / "raw"

        for d in [self.code_dir, self.docker_dir, self.events_dir,
                  self.reports_dir, self.raw_dir]:
            d.mkdir(parents=True, exist_ok=True)

        self.columns = columns or []
        self.results: Dict[str, List[TreatmentResult]] = {}
        self.metadata: Dict[str, Any] = {
            "experiment_id": self.experiment_id,
            "started_at": datetime.now().isoformat(),
            "treatments": [],
        }

    def add_result(self, treatment_name: str, result: TreatmentResult):
        """Add a treatment result."""
        if treatment_name not in self.results:
            self.results[treatment_name] = []
            self.metadata["treatments"].append(treatment_name)
        self.results[treatment_name].append(result)

    def _get_all_columns(self) -> List[ReportColumn]:
        """Get all columns: custom first, then defaults."""
        defaults = default_columns()
        pass_col = defaults[0]
        metric_cols = defaults[1:]
        return [pass_col] + self.columns + metric_cols

    def generate_summary(self) -> str:
        """Generate markdown summary of experiment results."""
        lines = []
        lines.append("# Experiment Results Summary\n")
        lines.append(f"**Experiment ID:** `{self.experiment_id}`\n")
        lines.append(f"**Started:** {self.metadata['started_at']}\n")
        lines.append(f"**Completed:** {datetime.now().isoformat()}\n")
        lines.append("")

        columns = self._get_all_columns()
        has_reps = any(len(runs) > 1 for runs in self.results.values())

        col_names = ["Treatment"] + [c.name for c in columns]
        header = "| " + " | ".join(col_names) + " |"
        separator = "|" + "|".join("-" * (len(name) + 2) for name in col_names) + "|"

        if has_reps:
            lines.append("## Results (with Repetitions)\n")
        else:
            lines.append("## Results\n")

        lines.append(header)
        lines.append(separator)

        for name, runs in self.results.items():
            if has_reps:
                values = [c.get_aggregate(runs) for c in columns]
            else:
                values = [c.get_value(runs[0]) for c in columns]
            lines.append(f"| {name} | " + " | ".join(values) + " |")

        lines.append("")

        total_runs = sum(len(runs) for runs in self.results.values())
        total_passed = sum(sum(1 for r in runs if r.passed) for runs in self.results.values())

        lines.append("## Summary\n")
        lines.append(f"- **Total Runs:** {total_runs}")
        lines.append(f"- **Pass Rate:** {total_passed}/{total_runs} ({100*total_passed//total_runs if total_runs else 0}%)")
        lines.append("")

        failed_runs = [
            (name, r) for name, runs in self.results.items()
            for r in runs if not r.passed
        ]
        if failed_runs:
            lines.append("## Failed Runs\n")
            for name, r in failed_runs[:10]:
                lines.append(f"- **{name}:** {', '.join(r.checks_failed[:2])}")
            if len(failed_runs) > 10:
                lines.append(f"- ... and {len(failed_runs) - 10} more")

        return "\n".join(lines)

    def finalize(self):
        """Generate and save final summary."""
        summary = self.generate_summary()
        summary_path = self.base_dir / "summary.md"
        summary_path.write_text(summary)

        self.metadata["completed_at"] = datetime.now().isoformat()
        self.metadata["total_runs"] = sum(len(runs) for runs in self.results.values())
        self.metadata["total_passed"] = sum(
            sum(1 for r in runs if r.passed) for runs in self.results.values()
        )

        metadata_path = self.base_dir / "metadata.json"
        metadata_path.write_text(json.dumps(self.metadata, indent=2))

        print(f"\nExperiment results saved to: {self.base_dir}")
        print(f"Summary: {summary_path}")

        return summary_path


# =============================================================================
# PARALLEL SAVE HELPERS (for multiprocessing workers)
# =============================================================================

def save_events(base_dir: Path, treatment_name: str, rep: int, events: Dict[str, Any]):
    """Save events JSON."""
    events_dir = base_dir / "events"
    save_path = events_dir / f"{treatment_name.lower()}_rep{rep}.json"
    save_path.write_text(json.dumps(events, indent=2))
    return save_path


def save_raw(base_dir: Path, treatment_name: str, rep: int, stdout: str, stderr: str = None):
    """Save raw CLI output."""
    raw_dir = base_dir / "raw"
    stdout_path = raw_dir / f"{treatment_name.lower()}_rep{rep}_stdout.json"
    stdout_path.write_text(stdout)

    if stderr:
        stderr_path = raw_dir / f"{treatment_name.lower()}_rep{rep}_stderr.txt"
        stderr_path.write_text(stderr)


def save_report(base_dir: Path, treatment_name: str, rep: int, report: Dict[str, Any]):
    """Save treatment report."""
    reports_dir = base_dir / "reports"
    save_path = reports_dir / f"{treatment_name.lower()}_rep{rep}_report.json"
    save_path.write_text(json.dumps(report, indent=2))
    return save_path

File: scaffold/typescript/logging.ts
/**
 * Output parsing, event extraction, and experiment logging.
 */

import { existsSync, mkdirSync, writeFileSync, readdirSync, readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
export const PROJECT_ROOT = resolve(__dirname, "..", "..");
export const LOGS_DIR = join(PROJECT_ROOT, "logs");

// eslint-disable-next-line no-control-regex
const ANSI_ESCAPE = /\x1B(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])/g;
const NPM_NOISE = [/^npm warn exec.*$/gm, /^npm notice.*$/gm];

export function stripAnsi(text: string): string {
  return text.replace(ANSI_ESCAPE, "");
}

export function stripNpmNoise(text: string): string {
  let result = text;
  for (const pattern of NPM_NOISE) result = result.replace(pattern, "");
  return result.replace(/\n{3,}/g, "\n\n").trim();
}

export function cleanOutput(text: string): string {
  return stripNpmNoise(stripAnsi(text));
}

// =============================================================================
// OUTPUT PARSING
// =============================================================================

export interface ParsedOutput {
  messages: Record<string, unknown>[];
}

/** Parse stream-json output into structured data. */
export function parseOutput(stdout: string): ParsedOutput {
  if (!stdout) return { messages: [] };
  const messages: Record<string, unknown>[] = [];
  for (const line of stdout.trim().split("\n")) {
    try { messages.push(JSON.parse(line)); } catch { /* skip non-JSON */ }
  }
  return { messages };
}

// =============================================================================
// EVENT EXTRACTION
// =============================================================================

export interface ToolCall {
  tool: string;
  input: Record<string, unknown>;
  output?: string;
}

export interface Events {
  tool_calls: ToolCall[];
  files_read: string[];
  files_created: string[];
  files_modified: string[];
  commands_run: string[];
  skills_invoked: string[];
  duration_seconds: number | null;
  num_turns: number | null;
}

/**
 * Extract events (tool calls, files, etc.) from parsed output.
 */
export function extractEvents(parsed: ParsedOutput): Events {
  const events: Events = {
    tool_calls: [],
    files_read: [],
    files_created: [],
    files_modified: [],
    commands_run: [],
    skills_invoked: [],
    duration_seconds: null,
    num_turns: null,
  };

  // Map tool_use_id -> index in tool_calls list for matching outputs
  const toolIdToIndex: Record<string, number> = {};

  for (const msg of parsed.messages) {
    if (msg.type === "result") {
      const durationMs = msg.duration_ms as number | undefined;
      events.duration_seconds = durationMs ? durationMs / 1000 : null;
      events.num_turns = (msg.num_turns as number) ?? null;
    }

    if (msg.type === "assistant") {
      const message = msg.message as Record<string, unknown> | undefined;
      const content = (message?.content as unknown[]) || [];

      for (const item of content) {
        const itemObj = item as Record<string, unknown>;
        if (itemObj.type === "tool_use") {
          const tool = (itemObj.name as string) || "";
          const inp = (itemObj.input as Record<string, unknown>) || {};
          const toolId = itemObj.id as string | undefined;

          const toolCall: ToolCall = { tool, input: inp };
          if (toolId) {
            toolIdToIndex[toolId] = events.tool_calls.length;
          }
          events.tool_calls.push(toolCall);

          const path = (inp.file_path as string) || "";
          if (tool === "Read" && path) {
            events.files_read.push(path);
          } else if (tool === "Write" && path) {
            events.files_created.push(path);
          } else if (tool === "Edit" && path) {
            events.files_modified.push(path);
          } else if (tool === "Bash" && inp.command) {
            events.commands_run.push(inp.command as string);
          } else if (tool === "Skill" && inp.skill) {
            events.skills_invoked.push(inp.skill as string);
          }
        }
      }
    }

    // Capture tool results and match to their tool_use calls
    if (msg.type === "user") {
      const message = msg.message as Record<string, unknown> | undefined;
      const content = (message?.content as unknown[]) || [];

      for (const item of content) {
        const itemObj = item as Record<string, unknown>;
        if (itemObj.type === "tool_result") {
          const toolUseId = itemObj.tool_use_id as string | undefined;
          if (toolUseId && toolUseId in toolIdToIndex) {
            const idx = toolIdToIndex[toolUseId];
            let outputContent = itemObj.content;

            if (Array.isArray(outputContent)) {
              outputContent = outputContent
                .map((c) => {
                  if (typeof c === "object" && c !== null && "text" in c) {
                    return (c as Record<string, unknown>).text;
                  }
                  return String(c);
                })
                .join(" ");
            }

            events.tool_calls[idx].output = String(outputContent);
          }
        }
      }
    }
  }

  return events;
}

// =============================================================================
// TREATMENT RESULT
// =============================================================================

export interface EventsSummary {
  duration_seconds?: number | null;
  num_turns?: number | null;
  tool_calls?: number;
  files_created?: string[];
  skills_invoked?: string[];
}

export interface TreatmentResult {
  name: string;
  passed: boolean;
  checks_passed: string[];
  checks_failed: string[];
  events_summary: EventsSummary;
  run_id: string;
}

export function createTreatmentResult(
  name: string,
  checksPassed: string[],
  checksFailed: string[],
  events: Events,
  runId = ""
): TreatmentResult {
  return {
    name,
    passed: checksFailed.length === 0,
    checks_passed: checksPassed,
    checks_failed: checksFailed,
    events_summary: {
      duration_seconds: events.duration_seconds,
      num_turns: events.num_turns,
      tool_calls: events.tool_calls.length,
    },
    run_id: runId,
  };
}

export function hasCheck(result: TreatmentResult, pattern: string): boolean {
  return result.checks_passed.some((c) => c.includes(pattern));
}

export function hasFailedCheck(result: TreatmentResult, pattern: string): boolean {
  return result.checks_failed.some((c) => c.includes(pattern));
}

// =============================================================================
// REPORT COLUMNS
// =============================================================================

export interface ReportColumn {
  name: string;
  description?: string;
  extract: (result: TreatmentResult) => string;
  aggregate?: (runs: TreatmentResult[]) => string;
}

export function boolColumn(name: string, pattern: string, description?: string): ReportColumn {
  return {
    name,
    description: description || `Checks if any passed check contains: \`${pattern}\``,
    extract: (r) => hasCheck(r, pattern) ? "Yes" : "No",
    aggregate: (runs) => `${runs.filter((r) => hasCheck(r, pattern)).length}/${runs.length}`,
  };
}

export function qualityColumn(name = "Quality"): ReportColumn {
  return {
    name,
    description: "Checks for [GOOD] or [LOW] quality rating from LLM evaluation",
    extract: (r) => {
      for (const c of r.checks_passed) {
        if (c.includes("[GOOD]")) return "Good";
        if (c.includes("[LOW]")) return "Low";
      }
      return "N/A";
    },
    aggregate: (runs) => `${runs.filter((r) => r.checks_passed.some((c) => c.includes("[GOOD]"))).length}/${runs.length}`,
  };
}

function formatChecks(r: TreatmentResult): string {
  const passed = r.checks_passed.length;
  const total = passed + r.checks_failed.length;
  const pct = total > 0 ? (passed / total) * 100 : 0;
  return `${passed}/${total} (${pct.toFixed(0)}%)`;
}

function aggregateChecks(runs: TreatmentResult[]): string {
  const totalPassed = runs.reduce((sum, r) => sum + r.checks_passed.length, 0);
  const totalChecks = runs.reduce(
    (sum, r) => sum + r.checks_passed.length + r.checks_failed.length,
    0
  );
  const pct = totalChecks > 0 ? (totalPassed / totalChecks) * 100 : 0;
  return `${totalPassed}/${totalChecks} (${pct.toFixed(0)}%)`;
}

function avg(values: (number | null | undefined)[], format: (n: number) => string): string {
  const filtered = values.filter((v): v is number => v != null);
  if (filtered.length === 0) return "N/A";
  return format(filtered.reduce((a, b) => a + b, 0) / filtered.length);
}

/**
 * Standard columns: Checks, Turns, Duration, Tools.
 */
export function defaultColumns(): ReportColumn[] {
  return [
    {
      name: "Checks",
      extract: formatChecks,
      aggregate: aggregateChecks,
    },
    {
      name: "Turns",
      extract: (r) => (r.events_summary.num_turns?.toString() ?? "N/A"),
      aggregate: (runs) =>
        avg(
          runs.map((r) => r.events_summary.num_turns),
          (n) => n.toFixed(0)
        ),
    },
    {
      name: "Duration",
      extract: (r) =>
        r.events_summary.duration_seconds != null
          ? `${r.events_summary.duration_seconds.toFixed(0)}s`
          : "N/A",
      aggregate: (runs) =>
        avg(
          runs.map((r) => r.events_summary.duration_seconds),
          (n) => `${n.toFixed(0)}s`
        ),
    },
    {
      name: "Tools",
      extract: (r) => r.events_summary.tool_calls?.toString() ?? "N/A",
      aggregate: (runs) =>
        avg(
          runs.map((r) => r.events_summary.tool_calls),
          (n) => n.toFixed(0)
        ),
    },
  ];
}

// =============================================================================
// EXPERIMENT LOGGER
// =============================================================================

export class ExperimentLogger {
  experimentId: string;
  name: string;
  timestamp: string;
  baseDir: string;
  eventsDir: string;
  reportsDir: string;
  rawDir: string;
  columns: ReportColumn[];
  results: Record<string, TreatmentResult[]>;
  metadata: {
    experiment_id: string;
    started_at: string;
    treatments: string[];
    completed_at?: string;
    total_runs?: number;
    total_passed?: number;
  };

  constructor(options: {
    experimentName?: string;
    columns?: ReportColumn[];
    experimentId?: string;
  } = {}) {
    const { experimentName, columns = [], experimentId } = options;

    if (experimentId) {
      // Join existing experiment
      this.experimentId = experimentId;
      this.name = experimentId.includes("_")
        ? experimentId.split("_").slice(0, -2).join("_")
        : experimentId;
      this.timestamp = experimentId.includes("_")
        ? experimentId.split("_").slice(-1)[0]
        : "";
    } else {
      // Create new experiment
      this.timestamp = new Date().toISOString().replace(/[-:]/g, "").slice(0, 15).replace("T", "_");
      this.name = experimentName || `experiment_${this.timestamp}`;
      this.experimentId = `${this.name}_${this.timestamp}`;
    }

    this.baseDir = join(LOGS_DIR, "experiments", this.experimentId);
    this.eventsDir = join(this.baseDir, "events");
    this.reportsDir = join(this.baseDir, "reports");
    this.rawDir = join(this.baseDir, "raw");

    // Create directories
    for (const dir of [this.eventsDir, this.reportsDir, this.rawDir]) {
      mkdirSync(dir, { recursive: true });
    }

    this.columns = columns;
    this.results = {};
    this.metadata = {
      experiment_id: this.experimentId,
      started_at: new Date().toISOString(),
      treatments: [],
    };
  }

  addResult(treatmentName: string, result: TreatmentResult): void {
    if (!(treatmentName in this.results)) {
      this.results[treatmentName] = [];
      this.metadata.treatments.push(treatmentName);
    }
    this.results[treatmentName].push(result);
  }

  private getAllColumns(): ReportColumn[] {
    const defaults = defaultColumns();
    const passCol = defaults[0];
    const metricCols = defaults.slice(1);
    return [passCol, ...this.columns, ...metricCols];
  }

  generateSummary(): string {
    const lines: string[] = [];
    lines.push("# Experiment Results Summary\n");
    lines.push(`**Experiment ID:** \`${this.experimentId}\`\n`);
    lines.push(`**Started:** ${this.metadata.started_at}\n`);
    lines.push(`**Completed:** ${new Date().toISOString()}\n`);
    lines.push("");

    const columns = this.getAllColumns();
    const hasReps = Object.values(this.results).some((runs) => runs.length > 1);

    // Column definitions section
    const customCols = this.columns.filter((c) => c.description);
    if (customCols.length > 0) {
      lines.push("## Column Definitions\n");
      for (const col of customCols) {
        lines.push(`- **${col.name}**: ${col.description}`);
      }
      lines.push("");
    }

    const colNames = ["Treatment", ...columns.map((c) => c.name)];
    const header = "| " + colNames.join(" | ") + " |";
    const separator = "|" + colNames.map((n) => "-".repeat(n.length + 2)).join("|") + "|";

    lines.push(hasReps ? "## Results (with Repetitions)\n" : "## Results\n");
    lines.push(header);
    lines.push(separator);

    for (const [name, runs] of Object.entries(this.results)) {
      const values = columns.map((c) =>
        hasReps && c.aggregate ? c.aggregate(runs) : c.extract(runs[0])
      );
      lines.push(`| ${name} | ${values.join(" | ")} |`);
    }

    lines.push("");

    const totalRuns = Object.values(this.results).reduce((sum, runs) => sum + runs.length, 0);
    const totalChecksPassed = Object.values(this.results).reduce(
      (sum, runs) => sum + runs.reduce((s, r) => s + r.checks_passed.length, 0),
      0
    );
    const totalChecks = Object.values(this.results).reduce(
      (sum, runs) =>
        sum + runs.reduce((s, r) => s + r.checks_passed.length + r.checks_failed.length, 0),
      0
    );
    const checkPct = totalChecks > 0 ? (totalChecksPassed / totalChecks) * 100 : 0;

    lines.push("## Summary\n");
    lines.push(`- **Total Runs:** ${totalRuns}`);
    lines.push(`- **Checks Passed:** ${totalChecksPassed}/${totalChecks} (${checkPct.toFixed(1)}%)`);
    lines.push("");

    // Detailed per-treatment breakdown
    lines.push("## Treatment Details\n");
    for (const [name, runs] of Object.entries(this.results)) {
      const treatmentPassed = runs.reduce((sum, r) => sum + r.checks_passed.length, 0);
      const treatmentTotal = runs.reduce(
        (sum, r) => sum + r.checks_passed.length + r.checks_failed.length,
        0
      );
      const treatmentPct = treatmentTotal > 0 ? (treatmentPassed / treatmentTotal) * 100 : 0;
      lines.push(
        `### ${name} (${treatmentPassed}/${treatmentTotal} checks, ${treatmentPct.toFixed(0)}%)\n`
      );

      for (let i = 0; i < runs.length; i++) {
        const r = runs[i];
        const runLabel = hasReps ? `Run ${i + 1}` : "Result";
        const runPassed = r.checks_passed.length;
        const runTotal = runPassed + r.checks_failed.length;
        const runPct = runTotal > 0 ? (runPassed / runTotal) * 100 : 0;
        const runIdStr = r.run_id ? ` (run_id: ${r.run_id})` : "";
        lines.push(`**${runLabel}:** ${runPassed}/${runTotal} checks (${runPct.toFixed(0)}%)${runIdStr}`);

        // Show metrics
        const metrics: string[] = [];
        if (r.events_summary.num_turns != null) {
          metrics.push(`Turns: ${r.events_summary.num_turns}`);
        }
        if (r.events_summary.duration_seconds != null) {
          metrics.push(`Duration: ${r.events_summary.duration_seconds.toFixed(0)}s`);
        }
        if (r.events_summary.tool_calls != null) {
          metrics.push(`Tool calls: ${r.events_summary.tool_calls}`);
        }
        if (metrics.length > 0) {
          lines.push(`- Metrics: ${metrics.join(", ")}`);
        }

        // Show all passed checks
        if (r.checks_passed.length > 0) {
          lines.push(`- Passed checks (${r.checks_passed.length}):`);
          for (const check of r.checks_passed) {
            lines.push(`  - ${check}`);
          }
        }

        // Show all failed checks
        if (r.checks_failed.length > 0) {
          lines.push(`- Failed checks (${r.checks_failed.length}):`);
          for (const check of r.checks_failed) {
            lines.push(`  - ${check}`);
          }
        }

        lines.push("");
      }
    }

    return lines.join("\n");
  }

  finalize(): string {
    const summary = this.generateSummary();
    const summaryPath = join(this.baseDir, "summary.md");
    writeFileSync(summaryPath, summary);

    this.metadata.completed_at = new Date().toISOString();
    this.metadata.total_runs = Object.values(this.results).reduce(
      (sum, runs) => sum + runs.length,
      0
    );
    this.metadata.total_passed = Object.values(this.results).reduce(
      (sum, runs) => sum + runs.filter((r) => r.passed).length,
      0
    );

    const metadataPath = join(this.baseDir, "metadata.json");
    writeFileSync(metadataPath, JSON.stringify(this.metadata, null, 2));

    console.log(`\nExperiment results saved to: ${this.baseDir}`);
    console.log(`Summary: ${summaryPath}`);

    return summaryPath;
  }
}

// =============================================================================
// PARALLEL SAVE HELPERS
// =============================================================================

export function saveEvents(baseDir: string, treatmentName: string, rep: number, events: Events): string {
  const dir = join(baseDir, "events");
  mkdirSync(dir, { recursive: true });
  const path = join(dir, `${treatmentName.toLowerCase()}_rep${rep}.json`);
  writeFileSync(path, JSON.stringify(events, null, 2));
  return path;
}

export function saveRaw(baseDir: string, treatmentName: string, rep: number, stdout: string, stderr?: string): void {
  const dir = join(baseDir, "raw");
  mkdirSync(dir, { recursive: true });
  writeFileSync(join(dir, `${treatmentName.toLowerCase()}_rep${rep}_stdout.json`), stdout);
  if (stderr) writeFileSync(join(dir, `${treatmentName.toLowerCase()}_rep${rep}_stderr.txt`), stderr);
}

export function saveReport(baseDir: string, treatmentName: string, rep: number, report: Record<string, unknown>): string {
  const dir = join(baseDir, "reports");
  mkdirSync(dir, { recursive: true });
  const path = join(dir, `${treatmentName.toLowerCase()}_rep${rep}_report.json`);
  writeFileSync(path, JSON.stringify(report, null, 2));
  return path;
}

export function loadResultsFromReports(reportsDir: string): TreatmentResult[] {
  if (!existsSync(reportsDir)) return [];
  const results: TreatmentResult[] = [];

  for (const file of readdirSync(reportsDir).filter((f) => f.endsWith(".json")).sort()) {
    try {
      const report = JSON.parse(readFileSync(join(reportsDir, file), "utf8")) as Record<string, unknown>;
      results.push({
        name: (report.name as string) || "unknown",
        passed: (report.passed as boolean) ?? false,
        checks_passed: (report.checks_passed as string[]) || [],
        checks_failed: (report.checks_failed as string[]) || [],
        events_summary: (report.events_summary as EventsSummary) || {},
        run_id: (report.run_id as string) || "",
      });
    } catch { /* skip invalid */ }
  }
  return results;
}
